src/http/web_socket.cr

```diff
@@ -100,52 +110,101 @@ class HTTP::WebSocket
+    end
+
+    def text
+      @data.as(String)
+    end
+
+    def binary
+      @data.as(Bytes)
+    end
+  end
+
+  private def handle_opcode(message)
+    case message.opcode
+    when .close?
+      close(message.text) unless closed?
+      @close_reason = message.text
+    when .ping?
+      pong(message.text) unless closed?
+    end
+
+    if message.opcode.text? || message.opcode.binary?
+      return true
+    else
+      return false
+    end
+  end
+
+  # Waits until a full String or Bytes message is received on the socket and returns it.
+  # The data can be accessed in the `message` property of the returned struct.
+  #
+  # This method will respond to PING frames with PONG and close when a CLOSE frame is received.
+  # Use `#receive_raw` if you don't want that.
+  def receive
+    loop do
+      data = receive_raw
+
+      if handle_opcode(receive_raw)
+        return data
+      end
+    end
+  end
+
+  def receive_raw!
     loop do
       begin
         info = @ws.receive(@buffer)
       rescue IO::EOFError
-        @on_close.try &.call("")
+        return Message.new(:close, "")
         break
       end
 
+      @current_message.write @buffer[0, info.size]
+      next unless info.final
+
       case info.opcode
-      when Protocol::Opcode::PING
-        @current_message.write @buffer[0, info.size]
-        if info.final
-          message = @current_message.to_s
-          @on_ping.try &.call(message)
-          pong(message) unless closed?
-          @current_message.clear
-        end
-      when Protocol::Opcode::PONG
-        @current_message.write @buffer[0, info.size]
-        if info.final
-          @on_pong.try &.call(@current_message.to_s)
-          @current_message.clear
-        end
-      when Protocol::Opcode::TEXT
-        @current_message.write @buffer[0, info.size]
-        if info.final
-          @on_message.try &.call(@current_message.to_s)
-          @current_message.clear
-        end
-      when Protocol::Opcode::BINARY
-        @current_message.write @buffer[0, info.size]
-        if info.final
-          @on_binary.try &.call(@current_message.to_slice)
-          @current_message.clear
-        end
-      when Protocol::Opcode::CLOSE
-        @current_message.write @buffer[0, info.size]
-        if info.final
-          message = @current_message.to_s
-          @on_close.try &.call(message)
-          close(message) unless closed?
-          @current_message.clear
-          break
-        end
+      when .text?
+        message = Message.new(:text, @current_message.to_s)
+      when .binary?
+        message = Message.new(:binary, @current_message.to_slice)
+      when .close?
+        message = Message.new(:close, @current_message.to_s)
+      when .ping?
+        message = Message.new(:ping, @current_message.to_s)
+      when .pong?
+        message = Message.new(:pong, @current_message.to_s)
+      end
+
+      @current_message.clear
+      return message.not_nil!
```

Comment:
I'd prefer a raising `else` branch (which would remove `Nil` type from `message`) instead of calling `not_nil!`.